python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,154 @@
+        else:
+            (
+                self.storage_filesystem,
+                self.storage_fs_path,
+            ) = get_fs_and_path(self.storage_path)
+
+        self.syncer: Optional[Syncer] = (
+            None
+            if self.storage_path == self.storage_cache_path
+            else _DefaultSyncer(
+                sync_period=self.sync_config.sync_period,
+                sync_timeout=self.sync_config.sync_timeout,
+                storage_filesystem=self.storage_filesystem,
+            )
+        )
+
+        # TODO(justinvyu): This is needed for the legacy Trainable syncing code.
+        # Remove this after syncer argument is removed from SyncConfig.
+        self.sync_config.syncer = self.syncer
+
+        self._create_validation_file()
+        self._check_validation_file()
+
+    def __str__(self):
+        attrs = [
+            "storage_path",
+            "storage_cache_path",
+            "storage_filesystem",
+            "storage_fs_path",
+            "experiment_dir_name",
+            "trial_dir_name",
+        ]
+        attr_str = "\n".join([f"  {attr}={getattr(self, attr)}" for attr in attrs])
+        return f"StorageContext<\n{attr_str}\n>"
+
+    def _create_validation_file(self):
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        self.storage_filesystem.create_dir(self.experiment_fs_path)
+        with self.storage_filesystem.open_output_stream(valid_file):
+            pass
+
+    def _check_validation_file(self):
+        valid_file = os.path.join(self.experiment_fs_path, ".validate_storage_marker")
+        valid = self.storage_filesystem.get_file_info([valid_file])[0]
+        if valid.type == pyarrow.fs.FileType.NotFound:
+            raise RuntimeError(
+                f"Unable to set up cluster storage at storage_path={self.storage_path}"
+                "\nCheck that all nodes in the cluster have read/write access "
+                "to the configured storage path."
+            )
+
+    @property
+    def experiment_fs_path(self) -> str:
+        return os.path.join(self.storage_fs_path, self.experiment_dir_name)
+
+    @property
+    def experiment_cache_path(self) -> str:
+        return os.path.join(self.storage_cache_path, self.experiment_dir_name)
+
+    @property
+    def trial_fs_path(self) -> str:
+        if self.trial_dir_name is None:
+            raise RuntimeError(
+                "Should not access `trial_fs_path` without setting `trial_dir_name`"
+            )
+        return os.path.join(self.experiment_fs_path, self.trial_dir_name)
+
+    def construct_checkpoint_fs_path(self, checkpoint_dir_name: str) -> str:
+        return os.path.join(self.trial_fs_path, checkpoint_dir_name)
+
+
+_storage_context: Optional[StorageContext] = None
+
+
+def init_shared_storage_context(storage_context: StorageContext):
+    global _storage_context
+    _storage_context = storage_context
+
+
+def get_storage_context() -> StorageContext:
```

Comment:
Let's document when this is available? E.g., within the trial driver, trial workers, and tune driver processes?

Reply:
Btw, if we are making StorageContext a process singleton, should we avoid passing it around as a class/method argument and get it from the singleton only?

Reply:
I made this a singleton for a later PR that passes the storage context from the trainable to the train worker ([see here](https://github.com/ray-project/ray/pull/37696/files#diff-45f9db5c2d16fb2f2ce2458f7a99cb4e5c0be43c4c2ce65b9c710f2e526f5260R447-R448)).

In the current PR, I don't actually create a global storage context, since each trial has a different `trial_dir_name`. I can clarify this with a comment, saying that the singleton is only created / accessible globally on the trial driver process (Trainable).

Reply:
Oh I see, that makes sense. Indeed we should clarify the singleton is only valid on the trials.